perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/RegistrarManagerImpl.java

```diff
@@ -2867,18 +2874,40 @@ private Application getApplicationById(int appId) {
 		} else {
 			parsedName = new HashMap<>();
 		}
-		// try to fill if initial parsing failed -> returned null
-		if (parsedName.get("firstName") == null) {
-			parsedName.put("firstName", federValues.get(shibFirstNameVar));
-		}
-		if (parsedName.get("lastName") == null) {
-			parsedName.put("lastName", federValues.get(shibLastNameVar));
+		// if the idp provided first name or last name, always use it
+		String fedFirstName = federValues.get(shibFirstNameVar);
+		String fedLastName = federValues.get(shibLastNameVar);
+
+		setIfNonEmpty(parsedName, fedFirstName, "firstName");
+		setIfNonEmpty(parsedName, federValues.get(shibLastNameVar), "lastName");
```

Comment:
You probably wanted to use ``fedLastName`` variable instead of repeating ``federValues.get(...)``.

Reply:
Yes, I have fixed it.